### PR TITLE
Add benchmark workflow

### PR DESCRIPTION
**Pull request by michelle-clayton-work**
Implement a workflow that runs benchmarks on different versions of go and openssl. This will run for every PR. Verified that it will catch an alloc increase: 
- https://github.com/golang-fips/openssl/pull/416

Posts results in the comments with links pointing to the specific job. Since comments have a character limit, the results will just show top regressions and shoudl show test failures.

- https://github.com/golang-fips/openssl/issues/408

### CODE REVIEW

File: .github/workflows/benchmark.yml [hunk 1]
```diff
@@ -0,0 +1,373 @@
...
+              core.setOutput('pr-number', '');
+            }
+
+  bench-ubuntu:
+    needs: setup
+    strategy:
+      fail-fast: false
+      matrix:
+        go-version: [1.25.x, 1.26.x]
+        openssl-version: [3.0.1, 3.0.13, 3.1.5, 3.2.1, 3.3.0, 3.3.1, 3.4.0, 3.5.0]
```
Comment (qmuntal) [inline .github/workflows/benchmark.yml]: Too many versions, just run one benchmark per minor version.

File: .github/workflows/benchmark.yml [hunk 2]
```diff
@@ -0,0 +1,373 @@
...
+
+    - name: Install benchstat
+      run: go install golang.org/x/perf/cmd/benchstat@latest
+
+    - name: Run benchmarks (base)
+      working-directory: base
+      run: |
+        export GO_OPENSSL_VERSION_OVERRIDE=${{ matrix.openssl-version }}
+        export CGO_ENABLED=1
+        go test -run='^$' -bench=. -count=7 -benchmem -timeout 30m ./... 2>&1 | tee ../base.txt || true
```
Comment (qmuntal) [inline .github/workflows/benchmark.yml]: count=7 is useful for local benchmarks, where you done want to spend a lot of time waiting. On CI we can increase the count a bit. Let's say 10. Same applies to other bench runs.

File: .github/workflows/benchmark.yml [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+              });
+            }
+
+      - name: Check result
+        run: |
+          if [[ "${{ contains(needs.*.result, 'failure') }}" == "true" || "${{ contains(needs.*.result, 'cancelled') }}" == "true" ]]; then
+            echo "One or more benchmark jobs detected regressions or were cancelled."
+            exit 1
+          fi
+          echo "All benchmark jobs passed."
```
Comment (qmuntal) [inline .github/workflows/benchmark.yml]: Add a new line here.

File: .github/workflows/benchmark.yml [hunk 4]
```diff
@@ -0,0 +1,373 @@
...
+          echo "found=false" >> "$GITHUB_OUTPUT"
+        fi
+
+    - name: Compare benchmarks
+      id: benchstat
+      run: |
+        benchstat base.txt head.txt | tee benchstat.txt
+        THRESHOLD="5"
+        if awk -v threshold="$THRESHOLD" '
+          /^[[:space:]]*│.*(sec\/op|B\/op|allocs\/op)/ { header=1; next }
```
Comment (qmuntal) [inline .github/workflows/benchmark.yml]: I don't understand this regrex. But I'm ok if it works. Just one comment: it is OK to regress a bit on cpu time given that we don't have a stable machine. On the other hand, allocations shouldn't regress, not even a bit, as they don't depend on the host performance.

**Review by qmuntal — CHANGES_REQUESTED** [review]
Looks good (with some nits)!

**Review by qmuntal — COMMENTED** [review]
Image

Green benchmarks have an expand arrow but there is nothing to expand. Can we avoid it?

File: .github/workflows/benchmark.yml (hunk 4 above)
Comment (michelle-clayton-work) [reply .github/workflows/benchmark.yml]: For clarity added a small comment and the longer description is:

Here's what each line does:

/│.*(B\/op|allocs\/op)/ — Detects an allocation header row → sets header="alloc"
/│.*sec\/op/ — Detects a timing header row → sets header="time"
/^$/ — Blank line resets context (sections are separated by blanks)
match($0, /\+([0-9]+\.[0-9]+)% \(p=/, m) — Extracts the regression percentage from strings like +5.19% (p=0.001 n=10). The \+ means it only matches increases (decreases show -). m[1] captures just the number (5.19).
/[0-9].*[µm]/ — This is the unit filter, only for sec/op rows. It checks that the benchmark time contains µ (microseconds) or m (milliseconds), skipping n (nanoseconds). So 1.234µ matches but 18.31n doesn't.
The two threshold rules:

Allocations: any +X% (> 0%) → regression
CPU time: +X% where X ≥ 5, but only if the row's time is ≥ 1µs (the [µm] filter)
The [µm] filter exists because nanosecond-level benchmarks (like NewSHA256 at 18ns) show large percentage swings from normal CI noise, producing false positives.

Comment (michelle-clayton-work) [reply .github/workflows/benchmark.yml]: Also it should flag any alloc regression with `(m[1]+0 > 0)`

Comment (michelle-clayton-work) [reply .github/workflows/benchmark.yml]: But I will create a test branch off this and see if it correctly flags it

**Review by qmuntal — COMMENTED** [review]
I wonder if using golang.org/x/perf to parse the bench results using Go rather than bash+awk would make this code a bit easier to maintain and to test. Could you give it a try? If it's too much of a change then we can leave it for a follow-up PR.

File: .github/workflows/benchmark.yml [hunk 5]
```diff
@@ -0,0 +1,379 @@
...
+        # CPU time regressions (sec/op): >= 5% fails, but only for benchmarks >= 1µs
+        # (benchstat units: n=nano, µ=micro, m=milli; /[0-9].*[µm]/ skips ns-level noise).
+        if awk '
+          /^[[:space:]]*│.*(B\/op|allocs\/op)/ { header="alloc"; next }
+          /^[[:space:]]*│.*sec\/op/ { header="time"; next }
+          /^$/ { header="" }
+          header == "alloc" && match($0, /\+([0-9]+\.[0-9]+)% \(p=/, m) { if (m[1]+0 > 0) found=1 }
+          header == "time" && /[0-9].*[µm]/ && match($0, /\+([0-9]+\.[0-9]+)% \(p=/, m) { if (m[1]+0 >= 5) found=1 }
+          END { exit !found }
+        ' benchstat.txt; then
```
Comment (dagood) [inline .github/workflows/benchmark.yml]: A thought on this: we could move it to a tool written in Go for maintainer familiarity and perhaps a more structured approach.

*Potentially*, a Go tool could also use https://pkg.go.dev/golang.org/x/perf/benchmath etc. directly, rather than running https://github.com/golang/perf/blob/master/cmd/benchstat/main.go and parsing the intended-for-humans output.

Comment (dagood) [reply .github/workflows/benchmark.yml]: A middle ground: there is a `-format csv` that might help simplify machine parsing. (I haven't tried it. We'd probably want to run benchstat twice in this case, once for parsing and again for the comment content.)

Comment (dagood) [reply .github/workflows/benchmark.yml]: Oh, somehow I missed that @qmuntal already commented basically the same idea to use the package, sorry! 😄

**Comment by michelle-clayton-work** [comment]
@qmuntal @dagood let me know what you think of the current change. Using x/perf is not too complicated, but it means adding it to go.mod which I wasn't sure if that was ideal considering it is not actually needed for the code but just CI so kept it in a separate go.mod

File: .github/workflows/benchmark.yml [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+      working-directory: base
+      run: |
+        export CGO_ENABLED=1
+        go test -run='^$' -bench=. -count=10 -benchmem -timeout 60m ./... 2>&1 | tee ../base.txt || true
+
+    - name: Run benchmarks (head)
+      working-directory: head
+      run: |
+        export CGO_ENABLED=1
+        go test -run='^$' -bench=. -count=10 -benchmem -timeout 60m ./... 2>&1 | tee ../head.txt || true
```
Comment (dagood) [inline .github/workflows/benchmark.yml]: Seems like a few decent chunks of duplication that could be shared somehow between bench-azl3 and bench-ubuntu?

**Review by dagood — COMMENTED** [review]
Good approach! For the backend repos in the microsoft org, we can put the tool in microsoft/go-infra to share (as a normal cmd/ tool) without needing to split out the go.mod there.

**Comment by michelle-clayton-work** [comment]
Putting all the benchstat results in a PR comment would cause it to go over the word count limit. Probably need discussions on what information we want to be in the comment / how to truncate to stay within the limits but still have all the necessary information

File: .github/benchcheck/main.go [hunk 7]
```diff
@@ -0,0 +1,181 @@
+// benchcheck reads two Go benchmark result files and detects regressions.
+//
+// Usage: benchcheck [flags] base.txt head.txt
```
Comment (dagood) [inline .github/benchcheck/main.go]: Seems like this should show up in the `-h` output?

I like the `mkcgo` approach of having a `func usage() {` near the top so it's still easy to find when just reading code.

**Review by qmuntal — CHANGES_REQUESTED** [review]
This is looking great!

I'm still seeing too much bash/js in the benchmark.yml, though. Can we expand benchcheck to work better with github actions? For example, it could directly print all the output that is currently generated in the "📊 RESULTS" step. Also, it could Generate the `status.txt` that is currently generated in the "Save status" step. It could also build the final repo.

Think about benchcheck as the implementation of the whole benchmark workflow, being the bechmark.yml file just a coordinator that calls benchcheck with the appropiate parameters.

File: .github/workflows/benchmark.yml [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+    steps:
+    - name: Install build tools (Ubuntu)
+      if: matrix.platform == 'ubuntu'
+      run: sudo apt-get update && sudo apt-get install -y build-essential
+
+    - name: Install build tools (AZL3)
+      if: matrix.platform == 'azl3'
+      run: |
+        sed -i 's/repo_gpgcheck=1/repo_gpgcheck=0/' /etc/yum.repos.d/*.repo
+        tdnf install -y --nogpgcheck gcc binutils kernel-headers glibc-devel openssl-devel git tar ca-certificates
```
Comment (dagood) [inline .github/workflows/benchmark.yml]: Why is gpgcheck being turned off?

File: .github/workflows/benchmark.yml [hunk 9]
```diff
@@ -0,0 +1,343 @@
...
+      id: test-failures
+      run: |
+        extract_failures() {
+          local file=$1 prefix=$2
+          # Extract a compact crash/failure summary including the first goroutine's
+          # stack trace (the one that crashed). Stops at the first blank line after
+          # the crash to avoid dumping all goroutines.
+          awk '
+            # Build errors
+            /^# / { print; next }
```
Comment (dagood) [inline .github/workflows/benchmark.yml]: There's still a few decently large chunks of awk, shell, js, and jq scripts in here. I suppose for a dev-time workflow the downsides aren't all that bad (can't test it, less maintainer familiarity, lang tooling doesn't work as well embedded in yml). Maybe the benefit of having it all in a single file is actually worthwhile. It's mainly piping and filtering, at the end of the day--I do think that `benchcheck` was the most important bit to extract. Not sure if it's worth doing more, but seemed worth commenting on.

File: .github/workflows/benchmark.yml [hunk 10]
```diff
@@ -0,0 +1,343 @@
...
+            # Test failures
+            /^--- FAIL/ { print; next }
+            # Package FAIL lines
+            /^FAIL\t/ { print; next }
+            # Start of a crash: capture signal, panic, and first goroutine stack
+            /^SIG[A-Z]+:/ || /^panic:/ {
```
Comment (dagood) [inline .github/workflows/benchmark.yml]: Could use test `-json` or `test2json` to make this more structured using official tools. (Goal being that it's easier to deal with breaking changes and edge cases.)

File: .github/workflows/benchmark.yml [hunk 11]
```diff
@@ -0,0 +1,343 @@
...
+    - name: Install build tools (AZL3)
+      if: matrix.platform == 'azl3'
+      run: |
+        sed -i 's/repo_gpgcheck=1/repo_gpgcheck=0/' /etc/yum.repos.d/*.repo
+        tdnf install -y --nogpgcheck gcc binutils kernel-headers glibc-devel openssl-devel git tar ca-certificates
+
+    - name: Install OpenSSL
+      if: matrix.openssl-version != ''
+      run: |
+        git clone --depth 1 --filter=blob:none --sparse ${{ github.server_url }}/${{ github.repository }} _openssl_setup
```
Comment (dagood) [inline .github/workflows/benchmark.yml]: Is the goal here to specifically use the latest version of the `v2` branch, rather than `head` or `base`? I would think that `head` would be a good place to use the OpenSSL setup scripts from.

I think that `head` would let you e.g. perf test a new version of OpenSSL with new support added in `head`, so it might actually be better?

Comment (michelle-clayton-work) [reply .github/workflows/benchmark.yml]: head makes sense to me. Will update!

File: .github/workflows/benchmark.yml (hunk 8 above)
Comment (michelle-clayton-work) [reply .github/workflows/benchmark.yml]: I think it was just related to azl3 being shipped without the gpg keys

File: .github/benchcheck/check.go [hunk 12]
```diff
@@ -0,0 +1,295 @@
...
+}
+
+func cmdCheck(args []string) {
+	fs := flag.NewFlagSet("check", flag.ExitOnError)
+	timeThreshold := fs.Float64("time-threshold", 5, "minimum sec/op regression percentage to flag")
+	minTime := fs.Duration("min-time", time.Microsecond, "minimum base time for sec/op checks")
+	alpha := fs.Float64("alpha", 0.05, "significance level for statistical tests")
+	fs.Usage = func() {
+		fmt.Fprintf(os.Stderr, "usage: benchcheck check [flags] base.txt head.txt\n\nFlags:\n")
+		fs.PrintDefaults()
```
Comment (dagood) [inline .github/benchcheck/check.go:59]: It's unlikely that someone will read this program without the full context of benchmark.yml, but I think it would be nice to explain here that it'll be writing some files to disk, and generally what the files are for.

Also (and kind of alternatively), this program could write the results to stdout by default and accept output flags to put them in files for you. `-o-regressions regressions.txt` etc. This helps clarify the benchmark.yml file because you can ctrl-f the filename in the upload step and see where they're supposed to be coming from.

File: .github/workflows/benchmark.yml (hunk 8 above)
Comment (dagood) [reply .github/workflows/benchmark.yml]: Interesting... would it make sense to use `mcr.microsoft.com/oss/go/microsoft/golang:1.25-azurelinux3.0` instead, like the test workflow does? Might be able to remove this step entirely?

FWIW, I don't get any gpg problems when I use this `mcr.microsoft.com/azurelinux/base/core:3.0` image locally. Still seems odd. 😄

File: .github/benchcheck/check.go [hunk 13]
```diff
@@ -0,0 +1,295 @@
...
+		MinTime:       minTime.Seconds(),
+		Alpha:         *alpha,
+	}
+
+	basePath, headPath := fs.Arg(0), fs.Arg(1)
+
+	// Extract test failures from both files.
+	var failureLines []string
+	if lines, err := extractFailuresFromFile(basePath, "base: "); err == nil {
+		failureLines = append(failureLines, lines...)
```
Comment (dagood) [inline .github/benchcheck/check.go]: Is dropping non-nil errors intentional? A typical API for this that integrates error handling reasonably would be:

```go
	var failureLines []string
	if failureLines, err = appendFailuresFromFile(failureLines, basePath, "base: "); err != nil {
		// handle err
```

File: .github/benchcheck/check.go [hunk 14]
```diff
@@ -0,0 +1,295 @@
...
+			inCrash = true
+			lines = append(lines, prefix+line)
+		case inCrash && line == "":
+			inCrash = false
+			lines = append(lines, "")
+		case inCrash:
+			lines = append(lines, prefix+line)
+		}
+	}
+	return lines
```
Comment (dagood) [inline .github/benchcheck/check.go]: Missing scanner error handling when the scan is done (or failed)?

File: .github/benchcheck/check.go [hunk 15]
```diff
@@ -0,0 +1,295 @@
...
+	}
+}
+
+func writeStatus(path string, regression, failures bool) {
+	f, err := os.Create(path)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "writing %s: %v\n", path, err)
+		return
+	}
+	defer f.Close()
```
Comment (dagood) [inline .github/benchcheck/check.go]: Missing error handling. Consider `os.WriteFile` to simplify?

File: .github/benchcheck/report.go [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+	}
+
+	fmt.Print(output)
+}
+
+type jobStatus struct {
+	regression   bool
+	testFailures bool
+	benchError   bool
+	failed       bool // derived: any of the above
```
Comment (dagood) [inline .github/benchcheck/report.go:107]: Consider a method instead of repeating `s.failed = true` below?

File: .github/benchcheck/report.go [hunk 17]
```diff
@@ -0,0 +1,155 @@
...
+}
+
+func readJobStatus(path string) jobStatus {
+	data, err := os.ReadFile(path)
+	if err != nil {
+		// Missing status file means the job errored before writing it.
```
Comment (dagood) [inline .github/benchcheck/report.go]: Couldn't the error be something other than `errors.Is(err, os.ErrNotExist)`?

File: .github/benchcheck/report.go [hunk 18]
```diff
@@ -0,0 +1,155 @@
...
+		return jobStatus{failed: true, benchError: true}
+	}
+	var s jobStatus
+	for _, line := range strings.Split(string(data), "\n") {
+		switch {
+		case line == "regression=true":
```
Comment (dagood) [inline .github/benchcheck/report.go]: This isn't a *lot* of code, but I feel like changing the payload to JSON would still reduce it and standardize the serialization/deserialization.

File: .github/workflows/benchmark.yml [hunk 19]
```diff
@@ -235,64 +166,12 @@ jobs:
...
-                echo "</details>"
-                echo ""
-              else
-                echo ":white_check_mark: \`$label\` · [results]($job_url)"
-              fi
-            done
-          } > report.md
+          go build -C head/.github/benchcheck -o "$PWD/benchcheck" .
+          ./benchcheck report \
+            --run-url="$RUN_URL" \
```
Comment (dagood) [inline .github/workflows/benchmark.yml]: This works, but Go flags are usually written `-run-url`.

File: .github/benchcheck/check.go (hunk 12 above)
Comment (michelle-clayton-work) [reply .github/benchcheck/check.go:59]: added some -o-... flags!

File: .github/benchcheck/check.go (hunk 13 above)
Comment (michelle-clayton-work) [reply .github/benchcheck/check.go]: updated

File: .github/benchcheck/report.go (hunk 16 above)
Comment (michelle-clayton-work) [reply .github/benchcheck/report.go:107]: Added func (s *jobStatus) setFailed(field *bool)

File: .github/benchcheck/report.go (hunk 17 above)
Comment (michelle-clayton-work) [reply .github/benchcheck/report.go]: updated so  readJobStatus checks errors.Is(err, os.ErrNotExist) specifically — other errors get logged to stderr before returning the error status

File: .github/workflows/benchmark.yml (hunk 8 above)
Comment (michelle-clayton-work) [reply .github/workflows/benchmark.yml]: This is the error message I get -- I'm not sure what the resolution would be here

```
Run tdnf install -y openssl-devel
Loaded plugin: tdnfrepogpgcheck
Refreshing metadata for: 'Azure Linux Official Cloud Native 3.0 x86_64'
repo md signature check: No public key
Error: TDNFVerifySignature 2004
Plugin error: repogpgcheck plugin error: failed to verify signature

Error(2004) : Unknown error 404
Error: Failed to synchronize cache for repo 'Azure Linux Official Cloud Native 3.0 x86_64'
Disabling Repo: 'Azure Linux Official Cloud Native 3.0 x86_64'
Refreshing metadata for: 'Azure Linux Official Base 3.0 x86_64'
repo md signature check: No public key
Error: TDNFVerifySignature 2004
Plugin error: repogpgcheck plugin error: failed to verify signature

Error(2004) : Unknown error 404
Error: Failed to synchronize cache for repo 'Azure Linux Official Base 3.0 x86_64'
Disabling Repo: 'Azure Linux Official Base 3.0 x86_64'
Refreshing metadata for: 'Azure Linux Official Microsoft Open-Source 3.0 x86_64'
repo md signature check: No public key
Error: TDNFVerifySignature 2004
Plugin error: repogpgcheck plugin error: failed to verify signature

Error(2004) : Unknown error 404
Disabling Repo: 'Azure Linux Official Microsoft Open-Source 3.0 x86_64'
Refreshing metadata for: 'Azure Linux Official Microsoft Non-Open-Source 3.0 x86_64'
Error: Failed to synchronize cache for repo 'Azure Linux Official Microsoft Open-Source 3.0 x86_64'
repo md signature check: No public key
Error: TDNFVerifySignature 2004
Plugin error: repogpgcheck plugin error: failed to verify signature

Error(2004) : Unknown error 404
Error: Failed to synchronize cache for repo 'Azure Linux Official Microsoft Non-Open-Source 3.0 x86_64'
Disabling Repo: 'Azure Linux Official Microsoft Non-Open-Source 3.0 x86_64'
openssl-devel package not found or not installed
Error(1011) : No matching packages
```

Comment (dagood) [reply .github/workflows/benchmark.yml]: Ah, it's because GitHub Actions changes HOME, but the image isn't expecting it. See https://github.com/Azure/azure-cli/issues/29835. Repros locally if I include `-e "HOME=/github/home"`.

This worked for me: `export GNUPGHOME=/root/.gnupg`

Comment (dagood) [reply .github/workflows/benchmark.yml]: Tracking issue on AzL's side: https://github.com/microsoft/azurelinux/issues/2812

**Review by qmuntal — APPROVED** [review]
This seems good to go! We can always improve things in future PRs as we find improvement opportunities.

File: .github/benchcheck/check.go [hunk 20]
```diff
@@ -0,0 +1,314 @@
...
+func isAllocUnit(unit string) bool {
+	return unit == "B/op" || unit == "allocs/op"
+}
+
+func writeLines(path string, lines []string) {
+	data := []byte(strings.Join(lines, "\n") + "\n")
+	if len(lines) == 0 {
+		data = nil
+	}
+	if err := os.WriteFile(path, data, 0o644); err != nil {
```
Comment (dagood) [inline .github/benchcheck/check.go:292]: Consider adding a quick `os.MkdirAll(filepath.Dir(path))` before this so we don't at some point(s) in the future end up wasting a dev iteration on having to explicitly pre-create a directory in whatever script is wrapping this tool.

File: .github/benchcheck/check.go [hunk 21]
```diff
@@ -0,0 +1,314 @@
...
+		fmt.Fprintf(os.Stderr, "writing %s: %v\n", path, err)
```
Comment (dagood) [inline .github/benchcheck/check.go:293]: It strikes me as odd to only write out to stderr. I think funcs like this should almost always return the error (with additional context added in via fmt.Errorf) so the caller can decide what to do.

About handling overall, I think that a failure to write these files should end up causing the command to exit 1 rather than 0, right? These files aren't just supplemental, so I don't see a reason the command should behave as if it was a success.

(It is reasonable to let the command try to finish up what it can, though, and use `errors.Join` to accumulate.)

File: .github/benchcheck/report.go (hunk 16 above)
Comment (dagood) [reply .github/benchcheck/report.go:107]: What came to mind for me was a getter `func (s *jobStatus) failed() { return s.regression || s.testFailures || s.benchError }`, this is nifty though. 🙂

File: .github/benchcheck/check.go [hunk 22]
```diff
@@ -0,0 +1,314 @@
...
+	}
+}
+
+// Status is the machine-readable status written by check and read by report.
+type Status struct {
+	Regression     bool `json:"regression"`
+	TestFailures   bool `json:"test_failures"`
+	BenchmarkError bool `json:"benchmark_error"`
+}
```
Comment (dagood) [inline .github/benchcheck/check.go:302]: (Near-)Duplicate of `type jobStatus` defined in report.go?

File: .github/benchcheck/report.go [hunk 23]
```diff
@@ -0,0 +1,166 @@
...
+	}
+	if status.BenchmarkError {
+		s.setFailed(&s.benchError)
+	}
+	return s
+}
+
+func loadJobURLs(path string) map[string]string {
+	if path == "" {
+		return nil
```
Comment (dagood) [inline .github/benchcheck/report.go:144]: Does the caller really expect nil (not even an empty map) to be a possible output?

File: .github/benchcheck/report.go [hunk 24]
```diff
@@ -0,0 +1,166 @@
...
+			urls[parts[0]] = parts[1]
+		}
+	}
+	return urls
+}
+
+func readFileContent(path string) string {
+	data, err := os.ReadFile(path)
+	if err != nil {
+		return ""
```
Comment (dagood) [inline .github/benchcheck/report.go:163]: It doesn't seem safe to ignore all possible errors here. (Similar in `loadJobURLs`.)

File: .github/benchcheck/check.go [hunk 25]
```diff
@@ -0,0 +1,314 @@
...
+
+func cmdCheck(args []string) {
+	fs := flag.NewFlagSet("check", flag.ExitOnError)
+	timeThreshold := fs.Float64("time-threshold", 5, "minimum sec/op regression percentage to flag")
+	minTime := fs.Duration("min-time", time.Microsecond, "minimum base time for sec/op checks")
+	alpha := fs.Float64("alpha", 0.05, "significance level for statistical tests")
+	regressionsOut := fs.String("o-regressions", "regressions.txt", "output file for regression details")
+	failuresOut := fs.String("o-failures", "failures.txt", "output file for test failure details")
+	statusOut := fs.String("o-status", "status.json", "output file for machine-readable status (JSON)")
```
Comment (dagood) [inline .github/benchcheck/check.go:44]: I think these should default to `""` and thus not write files by default. This means the benchmark.yml is forced to specify them by name, making benchmark.yml clearer.

Right now, reading benchmark.yml, it appears that e.g. `status.json` is written to by `echo '{"regression":false,"test_failures":false,"benchmark_error":true}' > status.json` and stays that way until `Upload results` when it's uploaded.

File: .github/workflows/benchmark.yml [hunk 26]
```diff
@@ -0,0 +1,218 @@
...
+      if: matrix.openssl-version != ''
+      run: sudo sh head/scripts/openssl.sh ${{ matrix.openssl-version }}
+
+    - uses: actions/setup-go@4b73464bb391d4059bd26b0524d20df3927bd417 # v6.3.0
+      with:
+        go-version: ${{ matrix.go-version }}
+        cache-dependency-path: head/go.mod
+
+    - name: Install benchstat
+      run: go install golang.org/x/perf/cmd/benchstat@latest
```
Comment (dagood) [inline .github/workflows/benchmark.yml:96]: Consider putting this dependency (as a `tool`) in the benchcheck go.mod so `latest` can't float to a version that could somehow take advantage of `pull-requests: write`.

**Review by dagood — COMMENTED** [review]
I agree this is good to have--some suggestions for polishing it a bit.

File: .github/workflows/benchmark.yml (hunk 26 above)
Comment (michelle-clayton-work) [reply .github/workflows/benchmark.yml:96]: Sounds good! Will put this in a followup issue! Maybe I can fix it for the backends implementation
